primary_beam_correction.py

```diff
@@ -0,0 +1,280 @@
+    return wt_average
+
+
+def primary_beam_correction(beam_pattern, path):
+    '''
+    Correcting for the primary beam effects.
+    '''
+    true_image = read_fits(path)
+    tr_img = []
+    nterm = true_image.header['NTERM']
+    for j in range(nterm, true_image.header['NSPEC']+nterm):
+        tr_img.append(np.ravel(true_image.data[0, j, :, :]))
+    snr = []
+    pbc_image = []
+    for i in range(len((beam_pattern))):
+        # Getting the rms noise in each image (before primary beam correction)
+        # Getting all the values with attenuated flux of less than 10% of the peak
+        ind = np.argwhere(beam_pattern[i] <= 0.10)
+        if len(ind) > 0:
+            beam_pattern[i][ind] = np.nan
+        snr.append(standard_deviation(tr_img[i]))
+        # To correct the effect of the beam we divide by the beam.
+        ratio = tr_img[i] / beam_pattern[i]
+        pbc_image.append(ratio)
+    # Primary beam corrected (pbc) image
+    pbc_image = np.array(pbc_image)
+    snr = np.array(snr)
+    # Calculating a weighted average of the indidvidual frequency lane images
+    corr_image = weigted_average(pbc_image, snr)
+    corr_image = corr_image.reshape(true_image.data.shape[2], true_image.data.shape[3])
+    # Add new axis
+    corr_image = corr_image[np.newaxis, np.newaxis, :, :]
+    return corr_image
+
+
+def write_new_fits(pbc_image, path, outputFilename):
+    '''
+    Write out a new fits image with primary beam corrected continuum in its first  plane
+    '''
+    images = read_fits(path)
+    hdr = images.header
+    newhdr = hdr.copy()
+    # change the frequency plane keywords, we don't want multiple frequency #axes
+    newhdr['CTYPE3'] = 'FREQ'
+    newhdr['NAXIS3'] = 1
+    newhdr['CDELT3'] = 1.0
+    try:
+        if 'CLEANBMJ' in newhdr:
+            if newhdr['CLEANBMJ'] < 0:
+                # Checking CLEANBMAJ in the history 
+                newhdr['BMAJ'] = float(newhdr['HISTORY'][-2][20:30])
+                newhdr['BMIN'] = float(newhdr['HISTORY'][-2][36:48])
+                newhdr['BPA'] = float(newhdr['HISTORY'][-2][55:])
+            else:
+                # add in required beam keywords
+                newhdr['BMAJ'] = newhdr['CLEANBMJ']
+                newhdr['BMIN'] = newhdr['CLEANBMN']
+                newhdr['BPA'] = newhdr['CLEANBPA']
+        elif 'BMAJ' in newhdr:
+            newhdr['BMAJ'] = newhdr['CLEANBMJ']
+            newhdr['BMIN'] = newhdr['CLEANBMN']
+            newhdr['BPA'] = newhdr['CLEANBPA']
+    except Exception:
+        logging.error('Exception occurred, keywords not found', exc_info=True)
+
+    data_new = pbc_image[:, 0:1, :, :]  # the shape of data_new = (1,1, npix, npix)
+    new_hdu = fits.PrimaryHDU(header=newhdr, data=data_new)
+    new_hdu.writeto(outputFilename, overwrite=True)
+    return
+
+
+def config_logging(message):
+    logging.basicConfig(format='%(message)s', level=logging.INFO)
+    return logging.info(message)
+
```

Comment:
The way you have done the logging is a bit messed up. I added a comment about this earlier but it seems to be lost.

You shouldn't need to do the `basicConfig` every single time you write a log message. The way to do it is something like this:
```
def config_logging(level):
    logging.basicConfig(format='%(message)s', level=logging.INFO)
```

then in the code you just use `logging.info(...)` or `logging.error(...)` or whatever. That way you aren't calling an expensive method every single time you write a log message.

Reply:
I have corrected it.